### Dominant Eigenvalue
The dominant eigenvalue is the value with the largest absolute value. If $\lambda < 1$  or $\lambda > 1$ the divergence will take place in an oscillatory way. If $\lambda = 1$ then there is no change. Between a set of eigenvalues that all show divergence, the one with the largest $\lambda$ value will be the vector to diverge the fastest.

The vectors after $k$ timesteps should be in the or opposite direction of the dominant eigenvector.

Does this mean that the cosine of the that the vector producted after $k$ timesteps and the dominant eigenvector should be $1$?

In [1]:
import numpy as np
import matplotlib.pyplot as plt

In [2]:
A = np.array([[1, 0, -1], 
              [0, 1, 1], 
              [-1, 1, -3]])
A

array([[ 1,  0, -1],
       [ 0,  1,  1],
       [-1,  1, -3]])

In [3]:
np.linalg.matrix_power(A, 10)

array([[ 10962, -10961,  48682],
       [-10961,  10962, -48682],
       [ 48682, -48682, 216651]])

In [4]:
np.linalg.eig(A)

(array([-3.44948974,  1.        ,  1.44948974]),
 array([[ 2.14186495e-01, -7.07106781e-01, -6.73887339e-01],
        [-2.14186495e-01, -7.07106781e-01,  6.73887339e-01],
        [ 9.53020614e-01,  1.33504240e-17,  3.02905447e-01]]))

In [12]:
np.linalg.eig(A)[1][:, 0].reshape(3, 1)

array([[ 0.2141865 ],
       [-0.2141865 ],
       [ 0.95302061]])

In [9]:
A.dot(np.linalg.eig(A)[1][:, 0].reshape(3, 1))

# Matrix times a "vector", which gives another vector, however, 
# in this case, the input vector is special in that it's an eigenvector.

array([[-0.73883412],
       [ 0.73883412],
       [-3.28743483]])

In [7]:
-3.44948974 * np.linalg.eig(A)[1][:, 0].reshape(3, 1)

array([[-0.73883412],
       [ 0.73883412],
       [-3.28743483]])

$A \in R^{n \times n}$

In [ ]:
def cosine_ab(a, b):
    '''
    Computes the Cosine of the angle 
    between vectors a and b.
    '''
    num = np.dot(a, b)
    dem_a = np.linalg.norm(a)
    dem_b = np.linalg.norm(b)
    return num / (dem_a * dem_b)

In [ ]:
np.linalg.matrix_power(A, 10)[:,0]

In [ ]:
A @ A @ A @ [[1], [1], [1]]

In [ ]:
B = np.array([1., 0., 0.]).reshape(3, 1)
B

In [ ]:
type(B)

In [ ]:
C = [[1.], [0.], [0.]]
C

In [ ]:
type(C)

In [ ]:
#I get different values from what was noted in the lecture. 
D = np.linalg.matrix_power(A, 1000)

In [ ]:
D.dot(C)

In [ ]:
D @ C

In [ ]:
E = np.linalg.eig(A)
A

In [ ]:
np.linalg.matrix_power(A, 10)

In [ ]:
cosine_ab(E[1][:,0], np.linalg.matrix_power(A, 10)[:,0])

In [ ]:
np.linalg.matrix_power(A, 1000) @ [[1.], [0.], [0.]]
# x and z are same sign.
# Should be in same direction as the dominant eigenvector.

In [ ]:
np.linalg.matrix_power(A, 1001) @ [[1.], [0.], [0.]]
#x and z are different sign.

In [ ]:
np.linalg.matrix_power(A, 1002) @ [[1.], [0.], [0.]]
#x and z are different sign.

In [ ]:
np.linalg.matrix_power(A, 1003) @ [[1.], [0.], [0.]]
#x and z are different sign.

In [ ]:
np.linalg.matrix_power(A, 1004) @ [[1.], [0.], [0.]]
#x and z are same sign.

In [ ]:
np.linalg.matrix_power(A, 1005) @ [[1.], [0.], [0.]]
#x and z are different sign.

In [ ]:
def dom_eigenvalue(A):
    eigen_obj = np.linalg.eig(A)
    pos = np.argmax([abs(each) for each in eigen_obj[0]])
    return eigen_obj[0][pos]

In [ ]:
print('Dominant Eigenvalue: {:.5f}'.format(dom_eigenvalue(A=A)))

In [ ]:
def cosine_ab(a, b):
    '''
    Computes the Cosine of the angle 
    between vectors a and b.
    '''
    num = np.dot(a, b)
    dem_a = np.linalg.norm(a)
    dem_b = np.linalg.norm(b)
    return num / (dem_a * dem_b)

In [ ]:
A1 = np.linalg.matrix_power(A, 1001) 
B1 = np.array([1., 0., 0.]).reshape(3,1)

In [ ]:
A1

In [ ]:
product = A1.dot(B1)
product.reshape(1, 3)

In [ ]:
eigen_vector = E[1][:,0]
eigen_vector

In [ ]:
#I expected this number to be either -1 or 1 since it has to be in the same direction as the eigenvector.
cosine_ab(product.reshape(1, 3), eigen_vector)

### Exercise 5.12

In [ ]:
A = np.array([[1, -1, 0], 
              [-1, -3, 1], 
              [0, 1, 1]])

In [ ]:
np.linalg.eig(A)

The dominant eigenvalue is $|\lambda| = 3.45$. Given the fact that $|\lambda| > 1$, the dominant eigenvectors are diverging to infinity (i.e., growing).

In [ ]:
#Dominant Eigenvector.
np.linalg.eig(A)[1][:,0].reshape(3, 1)

In [ ]:
np.linalg.matrix_power(A, 1002)

In [ ]:
np.linalg.matrix_power(A, 1003)

In [ ]:
np.linalg.matrix_power(A, 1004)

### Exercise 5.11

In [ ]:
def model(x_init, y_init, itr):
    
    xresults = [x_init]
    yresults = [y_init]
    
    for each in range(0, itr):
        x_new = xresults[-1] + yresults[-1]
        y_new = xresults[-1] 
        
        xresults.append(x_new)
        yresults.append(y_new)
    
    results = {}
    
    results['xresults'] = xresults
    results['yresults'] = yresults
    
    return results

In [ ]:
list(np.arange(-2, 2, 0.5))

In [ ]:
run = model(x_init=-0.09, y_init=0.025, itr=30)
plt.plot(run['xresults'], run['yresults'])
plt.show()